### PR TITLE
Use a Semaphore for signaling in repo fetching

### PR DESCRIPTION
**Pull request by Wyverald**
I managed to reproduce some deadlocks during repo fetching with virtual worker threads. One notable trigger was some _other_ repo failing to fetch, which seems to cause Skyframe to try to interrupt other concurrent repo fetches. This _might_ be the cause for a deadlock where we submit a task to the worker executor service, but the task never starts running before it gets cancelled, which causes us to wait forever for a `DONE` signal that never comes. (The worker task puts a `DONE` signal in the queue in a `finally` block -- but we don't even enter the `try`.)

This PR improves the situation in various ways:

1. Instead of using a `SynchronousQueue` for the signal queue, we now use a Semaphore for signaling. Semaphores have the crucial property that releasing a permit (ie. incrementing the counter) does not block, and thus cannot be interrupted. This means that the worker thread can now reliably send signals the host thread, even when it's interrupted.

2. Instead of using two signals for `DONE` and `RESTART`, we just use the one semaphore for both signals, and rely on `workerFuture.isDone()` to tell whether the worker has finished or is waiting for a fresh Environment.

3. Instead of signaling `DONE` in a `finally` block, we now use a `ListenableFuture` and signal to the semaphore in the worker future's listener. This makes sure that the signaling is performed _after_ the worker future's status changes, and safeguards against the case where the submitted task never starts running before it gets cancelled.

4. Instead of waiting for a `DONE` signal (or, in the new setup, the signal semaphore) to make sure the worker thread has finished, we now hold on to the executor service, which offers a `close()` method that essentially uninterruptibly waits for any scheduled tasks to terminate, whether or not they have started running. (@justinhorvitz had suggested a similar idea before.) To make sure distinct repo fetches don't interfere with each other, we start a separate worker executor service for each repo fetch instead of making everyone share the same worker executor service. (This is recommended for virtual threads; see https://docs.oracle.com/en/java/javase/21/core/virtual-threads.html#GUID-C0FEE349-D998-4C9D-B032-E01D06BE55F2 for example.)

And because I now create a separate worker executor service for each repo fetch, it doesn't really make sense to use this for platform threads anymore. So setting `--experimental_worker_for_repo_fetching` to any but `off` will cause us to use virtual threads.

Related: https://github.com/bazelbuild/bazel/issues/22003

Fixes https://github.com/bazelbuild/bazel/issues/21712.

### CODE REVIEW

**Comment by Wyverald** [comment]
cc @meteorcloudy @coeuvre

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:102]: one thing I don't quite understand here: I tried to replace this method body with

```java
workerFuture = null;
workerExecutorService.shutdownNow();
```

Per the [Javadoc of `shutdownNow`](https://docs.oracle.com/en%2Fjava%2Fjavase%2F21%2Fdocs%2Fapi%2F%2F/java.base/java/util/concurrent/ExecutorService.html#shutdownNow()), this should have the same effect as the previous code, but for some reason using `shutdownNow()` here consistently causes the later call to `workerExecutorService.close()` on line 108 to block forever. This seems rather surprising, but I have no idea whether this is a bug in JRE or I'm holding it wrong, so I just reverted to the previous implementation.

If anyone has an idea what's happening here, please share.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:161]: +1

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:128]: Maybe this change also makes callig `inErrorBubblingForSkyFunctionsThatCanFullyRecoverFromErrors()` unnecessary?

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:82]: nit: maybe add the name of the repository being fetched to the name of the thread?

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:102]: While the Skyframe thread is blocking on line 108, what is the worker virtual thread doing? (jstack should be able to tell)

**Comment by lberki** [comment]
Now that platform threads are not used anymore, why do we even need an `ExecutorService` and monkey around with futures? I could imagine an approach where we'd directly store a virtual thread in `RepoFetchingSkyKeyComputeState` and we `join()` it in a suitably-placed `finally` block and in case an early exit is needed, we can call `Thread.interrupt()` on it.

Granted, that thread could still swallow interrupts but at least there would be a bit fewer moving parts: that way, no questions would arise about whether we are using the `ExecutorService` correctly and it doesn't bring anything useful to the table anyway because we only ever pass one task to it.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (coeuvre) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:102]: You want to use `jcmd ${pid} Thread.dump_to_file ${out_file}` to capture virtual thread stacks.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:161]: just FYI this is for the situation described on line 202, although I didn't actually test if not having this `catch` would crash us.

this is likely to become moot when we switch to a `Thread`.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:128]: I'd rather not poke that bear until absolutely necessary, since that one was also hard to reproduce.

**Comment by Wyverald** [comment]
> > Now that platform threads are not used anymore, why do we even need an `ExecutorService` and monkey around with futures?
> 
> that's a good point -- the executor service really has no value anymore, so I'll probably just switch to a `Thread`. The future is still nice as it encapsulates the result [success (a value), failure (an exception), cancellation etc] quite nicely, so I'll probably still use it, but just as a SettableFuture or something.

Okay, I gave up. The failed attempt is at https://github.com/bazelbuild/bazel/pull/22110 if anyone's interested. It turns out using an executor service for a single task is still valuable since it manages the lifecycle of the underlying thread (with some latches etc). Directly using a thread while also trying to communicate the result/exception back to the host thread turned out to be rather tricky; in particular, while trying to avoid a case where the worker thread dies without emitting a signal, I always fell into some other deadlock (see failed PR for comments).

It seems to me that there's always some tension between the signal queue and the actual lifecycle of the worker thread. The host thread needs to listen on the signal queue while _also_ watching out for the worker thread dying before emitting a signal. Even with the executor service approach, I'm not confident that we're free of deadlocks; for example, if the worker is interrupted due to memory pressure, the host will probably wait forever to take something from the signal queue until it gets interrupted by something.

It almost feels like we need some sort of channel select mechanism -- upon any of 1) worker interrupted, 2) host interrupted, 3) worker emits signal, we need to take some action.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:105]: If you allow me to nitpick a bit, how could one interrupt a carrier thread? My understanding is that carrier threads are OS-level functionality that's not surfaced in Java (you don't get a `Thread` object to represent the carrier thread, do you?) and interrupts are Java-level functionality which doesn't have an equivalent in the OS.

**Comment by lberki** [comment]
I think I have an explanation for the deadlock you found on #22110. you interrupt the worker thread, then the `InterruptedException` is caught in the `catch (Throwable e)` in line 160, then it'll try to put something on the signal queue in the finally clause, but there is no one there to get that thing from the signal queue, so it's a deadlock. 

I can imagine two solutions to this:

1. Don't use the `BlockingQueue` to report that the thread is done in case an interrupt is received. This would work because even if the worker thread wants to put something in the queue when the interrupt hits it, the interrupt will unblock it. So doing nothing in the `catch (InterruptedException e) {}` branch is safe. This probably only works if the only thing that can interrupt the worker thread is the Skyframe thread (Can a virtual thread get interrupted in any way other than by `Thread.interrupt()`? Maybe a SIGINT to its carrier thread, whatever it may be?)

2. Use a unlimited queue (so that it never blocks) instead of a `BlockingQueue`. This would get rid of the synchronization point between `.put()` and `.take()` and make the code more resilient to cases you haven't anticipated and you could e.g. simply interrupt the worker thread, then drain the queue until you get the "I am done" message. FWIW, we use this approach in quite a few places in the code base (look for `POISON_PILL`, all caps), although I'm not sure if there is bidirectional communication in any of these places.

I think I'd do (2) for the very practical reason that you have carefully orchestrating a`SynchronousQueue` and it didn't work, but in either case, I think thinking through all the possible sequence of events is not optional...

**Comment by lberki** [comment]
I take the "unbounded queue" idea back. The only case where a deadlock should happen is when one thread is waiting in `.put()` / `.take()` and the other doesn't ever get there and it looks like it's reasonably simple to avoid that: in particular, I think "poison pill" approach to signal from one side that it should not expect any more updates from the other is pretty sound and easy to grok.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:105]: No? Carrier threads are Java threads. Virtual threads are entirely implemented in the JRE, AFAIK, not OS-level functionality. There's even a `VirtualThread#carrierThread` field (https://github.com/openjdk/jdk21/blob/890adb6410dab4606a4f26a942aed02fb2f55387/src/java.base/share/classes/java/lang/VirtualThread.java#L131).

**Comment by lberki** [comment]
My apologies; instead of "`finally` clause", I should have said `catch (Throwable e)`.

I took your attempt and ran with it and created https://github.com/bazelbuild/bazel/pull/22139 . WDYT? At the time I'm writing this, `test_no_restarts_fetching_with_worker_thread` still fails but at least there are no deadlocks which is a good start :) By the time you wake up, I hope I'll be able to polish it up to code.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:203]: Why does this condition cause an ExecutionException? I that in the case of b/330892334, the symptom would be that the worked thread returns with `Success(null)`, not an exception.

Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:203]: why does which condition cause an ExecutionException? this is a `finally` branch.

Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:203]: b/330892334 . TBH I don't know what I was thinking when I wrote this comment :(

**Comment by justinhorvitz** [comment]
Since you have feedback from other reviewers, and I have no experience with virtual threads, I removed myself as a reviewer. Please let me know if you want me to take a look at anything in particular though.

**Comment by Wyverald** [comment]
I finally found a solution I'm happy with! Please see the updated PR description. This new approach using Semaphores is IMO pretty clean and easy to understand.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java]: update: I poked the bear. Fairly sure that this is fixed.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:149]: Why is this call necessary here? AFAIU after this is called, the only thing that could possibly ever happen to `RepoFetchingSkyKeyComputeState` is another `close()` call.

In addition the aesthetic issue, this creates an unnecessary `ExecutionService`.

**Review by lberki — COMMENTED** [review]
Out of all the solutions, I like this one the best. As appealing as my Thread-based solution is, it somehow always ends up being more complicated than necessary. This one arguably has a lot of moving parts, but at least not a lot of code, which is a win.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:149]: in the case of test_keep_going_weird_deadlock, the SkyFunction is called a second time. Similar to error bubbling, IIUC.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:181]: yes. which part do you want me to turn into a loop? I thought the self call was very clear about what it's doing.

(but in any case, we'll probably need to do something different here due to https://github.com/bazelbuild/bazel/issues/22206; I just wanted to do it in another PR)

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:134]: if the worker thread gets 'orphaned' while the host Skyframe thread is active, then the host would call `state.closeAndWaitForTermination()`, ensuring the worker thread is joined. But if it gets 'orphaned' when the host Skyframe thread is _inactive_ (ie. having returned `null` but not yet restarted), then you're right the worker thread would live on for an indefinite amount of time.

I added a clause to guard against that specific case (so when the Skyframe thread becomes active again, the first thing it does is to join the worker thread).

Also added a TODO to reconsider the ban on blocking work in `close()`...

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:192]: OMG another deadlock? Ordinarily, I'd say that it's up to you, this call should not happen frequently enough to be a problem and from the code readability side, it's a matter of taste, but #22206 is an argument for representing the state in a more obvious way (like #22215 )

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:134]: My $.02: can we not have another complicated dance? The documentation of `close()` doesn't say *why* it shouldn't perform any blocking operations and it looks silly that a stray thread is preferred to blocking a bit.

I'd much rather not have to think about what happens in odd cases likewhen the Skyframe evaluation is stopped due to an independent error (or a SIGINT) before the worker thread quits and another Bazel command is issued right after that, etc.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:150]: Huh, remind me again, why is the SkyFunction called a second time in `test_keep_going_weird_deadlock`?

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:192]: it's yet unclear whether it's a deadlock, or just a really memory-constrained environment making fetch extra slow.

> but https://github.com/bazelbuild/bazel/issues/22206 is an argument for representing the state in a more obvious way

I don't see how using a loop is "more obvious" than a self call, though? Also, is it okay if we push to a follow-up PR? I'd rather not debate the coding style when we might get rid of the "reattmpt the fetch" logic altogether.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:150]: because the host Skyframe thread returns `null`, but the worker has already died (in fact, the worker itself returned a `null` in the Future). Normally the worker should never return `null` -- on `valuesMissing` it would block for a fresh env and try again; but in this case, even the fresh env still returns `null` for a `getValue` call, so the worker just returns that `null`.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:134]: > The documentation of close() doesn't say why it shouldn't perform any blocking operations

Presumably because it can be called from any thread. It's a bit scary to perform arbitrary blocking operations in random threads. If we want to reverse the decision, I'd rather get the Core team's input, and I suggest that we do this in a follow-up instead of right away in this PR, which is getting a bit big.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:192]: Fair enough. Let me take a look at this one more time to see if I spot any flaws.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:134]: Well, but at the same time, it's pretty scary to have orphaned threads doing I/O. But then again, that's arguably not for this pull request; you do cancel the future, which is supposed to kill off the now-orphan thread very quickly.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:139]: nit: doesn't `shutdownNow()` cancel all pending work?

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:150]: Thanks! Please add a comment to explain this scenario. If I haven't figured this out by myself after staring at this problem for so much time, people coming in fresh have no chance.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (lberki) [inline src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:198]: I assume that in this case, the return value of this SkyFunction will be null, but how? How does the worker thread guarantee that in that case, it will return null?

**Comment by lberki** [comment]
I added some comments, but I think that they are just asking for reassurance and the code works as it is today.

I propose the following course of action:

1. I approve this pull request
2. @Wyverald takes a look at my comments to see if they indicate and actual issues
3. If not, he merges it
4. I drop #22139 since it's strictly inferior to #22215
5. We continue iterating on #22215

**Review by lberki — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkRepositoryFunction.java:198]: > I assume that in this case, the return value of this SkyFunction will be null, but how?

It's not null IIUC. For b/330892334 it'll presumably throw on the newly-discovered already-eval'd failing `env.getValue()`; for #21238 it's error-bubbling, so it should throw some previously thrown exception (or "bubble up", whatever that may mean).

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:152]: done.

File: src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java
Comment (lberki) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:139]: TIL. Out of curiosity, did you experimentally verify this? It makes some weird sort of sense, but AFAIU each task an `ExecutorService` runs is wrapped in a future, so what would be the point of interrupting a task without cancelling the associated future?

Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/RepoFetchingSkyKeyComputeState.java:139]: I haven't experimentally verified this. But anyhow, the `ThreadPerTaskExecutor` does not hold on to the `Future`s it returns from its `submit()` method; also it does also have an `execute()` method that doesn't create a `Future` at all.

**Comment by ismell** [comment]
Was this cherry picked into 7.2.0rc2?

**Comment by lberki** [comment]
I thought it was but you're right -- it seems to not have been:

https://github.com/bazelbuild/bazel/compare/release-7.1.0...release-7.2.0rc2

*Very* good catch!

**Comment by lberki** [comment]
Nevermind, it's late here. The fix is de4d519ac52bd6400a016654eb85733ba16ac6f7 and was duly cherry-picked to 7.2.0rc2 as c6cabd8b45c46640f77190d08ba645e49f05a4e9. Please don't tell me that you still see this deadlock in rc2...

**Comment by ismell** [comment]
Oh good. I was just confirming since we just upgraded to 7.2.0rc2 and I wanted to make sure this bug fix landed.

Thanks!

**Comment by lberki** [comment]
🥁

**Comment by TeoMihuc** [comment]
Hello, was this added into 7.4.0?

**Comment by tjgq** [comment]
Yes, it's on 7.4.0 (7.4.0 incorporates all previous cherry-picks into 7.x.y).

**Comment by TeoMihuc** [comment]
thank you for your answer!

I have a weird situation which landed me here.

I am doing a bazel build inside a docker build RUN step, on a linux system - doing a cross-compilation from amd to arm using qemu. And bazel randomly gets stuck at a step saying 'no actions running', it stays in this state until I kill the build.

Do you have any recommendations? Should I open an issue for investigation/advice?

Thanks in advance!

**Comment by tjgq** [comment]
Can you capture a thread dump once Bazel appears to have reached the hung state (see https://bazel.build/rules/performance#hanging) and open a new issue with it? (Ideally, if you have a self-contained repro you can share it would be even better, but I understand that might be difficult.)

**Pull request closed without merging**
